doppelganger/doppelganger.go

```diff
@@ -0,0 +1,265 @@
+package doppelganger
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	eth2apiv1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"go.uber.org/zap"
+
+	"github.com/ssvlabs/ssv/logging"
+	"github.com/ssvlabs/ssv/logging/fields"
+	"github.com/ssvlabs/ssv/networkconfig"
+	"github.com/ssvlabs/ssv/operator/slotticker"
+	"github.com/ssvlabs/ssv/protocol/v2/types"
+)
+
+//go:generate mockgen -package=doppelganger -destination=./doppelganger_mock.go -source=./doppelganger.go
+
+// DefaultRemainingDetectionEpochs represents the initial number of epochs
+// a validator must pass without liveness detection before being considered safe to sign.
+const DefaultRemainingDetectionEpochs uint64 = 2
+
+// PermanentlyUnsafe is a special flag value used to mark a validator as permanently unsafe for signing.
+// It indicates that the validator was detected as live on another node and should not be trusted for signing.
+const PermanentlyUnsafe = ^uint64(0)
+
```

Comment:
since we have a low range of values (2/1/0) and the permanently unsafe. I think we should switch to using `int8` and `PermanentlyUnsafe` could be `-1`. this will make the data structure a bit lighter.

Reply:
Good point, but I’d prefer to keep it as uint64. The memory savings with int8 are negligible here, and switching might introduce unnecessary type conversions.

Reply:
if we decide to keep it uint64, I think it makes more sense it'll be using the `Epoch`, then we could use eth's `FarFutureEpoch` . I just don't like saving max int in mem many times.

Reply:
> if we decide to keep it uint64, I think it makes more sense it'll be using the `Epoch`, then we could use eth's `FarFutureEpoch` . I just don't like saving max int in mem many times.

sound reasonable. I will use `FarFutureEpoch`